File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.js
let fs = require("fs");
let _ = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

let [N, ...arr] = _;
N = Number(N);
arr = arr.map((v) => v.split(" ").map(Number));

const graph = Array.from({ length: N + 1 }, () => []);
const visited = Array.from({ length: N + 1 }).fill(0);
const head = Array.from({ length: N + 1 }, () => 0);
for (let [a, b] of arr) {
  graph[a].push(b);
  graph[b].push(a);
}

function DFS(node, s) {
  if (s === N) {
    return;
  } else {
    for (let next of graph[node]) {
      if (visited[next]) continue;
      visited[next] = 1;
      head[next] = node;
      DFS(next, s + 1);
    }
  }
}

DFS(1, 0);
let answer = "";
for (let i = 2; i < head.length; i++) {
  answer += head[i] + "\n";
}

console.log(answer);

File: 백준/Silver/11725. 트리의 부모 찾기/트리의 부모 찾기.py
import sys
import collections
import math
from collections import deque
import copy
import bisect
import itertools
import heapq

#sys.stdin = open("input.txt", "r")
input = sys.stdin.readline

N = int(input())
graph = [[] for _ in range(N+1)]
for _ in range(N-1):
    x,y = map(int,input().split())
    graph[x].append(y)
    graph[y].append(x)
    
parents = [1 for _ in range(N+1)]
visited = [0 for _ in range(N+1)]

def dfs(node):
    global parents, visited
    visited[node] = 1
    for nextNode in graph[node]:
        if visited[nextNode] == 1:
            continue
        parents[nextNode] = node
        dfs(nextNode)

def bfs(node):
    global parents, visited
    queue = deque()
    visited[node] = 1
    queue.append(node)
    
    while queue:
        node = queue.popleft()
        for nextNode in graph[node]:
            if visited[nextNode] == 1:
                continue
            parents[nextNode] = node
            visited[nextNode] = 1
            queue.append(nextNode)

for i in range(1,N+1):
    if visited[i] == 0:
        bfs(i)

for i in range(2,N+1):
    print(parents[i])